packages/fxa-auth-server/lib/routes/recovery-phone.ts

```diff
@@ -395,26 +394,112 @@ class RecoveryPhoneHandler {
 
     let success = false;
     try {
-      if (isSetup) {
-        // This is the initial setup case, where a user is validating an sms
-        // code on their phone for the first time. It does NOT impact the totp
-        // token's database state.
-        success = await this.recoveryPhoneService.confirmSetupCode(uid, code);
-      } else {
-        // This is a sign in attempt. This will check the code, and if valid, mark the
-        // session token verified. This session will have a security level that allows
-        // the user to remove totp devices.
-        success = await this.recoveryPhoneService.confirmCode(uid, code);
-
-        // Mark session as verified
-        if (success) {
-          await this.accountManager.verifySession(
+      // This is a sign in attempt. This will check the code, and if valid, mark the
+      // session token verified. This session will have a security level that allows
+      // the user to remove totp devices.
+      success = await this.recoveryPhoneService.confirmCode(uid, code);
+
+      // Mark session as verified
+      if (success) {
+        await this.accountManager.verifySession(
+          uid,
+          sessionTokenId,
+          VerificationMethods.sms2fa
+        );
+      }
+    } catch (error) {
+      if (error instanceof RecoveryPhoneNotEnabled) {
+        throw AppError.featureNotEnabled();
+      }
+
+      if (error instanceof RecoveryNumberAlreadyExistsError) {
+        throw AppError.recoveryPhoneNumberAlreadyExists();
+      }
+
+      if (error instanceof RecoveryPhoneRegistrationLimitReached) {
+        throw AppError.recoveryPhoneRegistrationLimitReached();
+      }
+
+      throw AppError.backendServiceFailure(
+        'RecoveryPhoneService',
+        'confirmCode',
+        { uid },
+        error
+      );
+    }
+
+    if (success) {
+      await this.glean.twoStepAuthPhoneCode.complete(request);
```

Comment:
Question: the same glean metric is/was emitted for both signin and setup. Should these be differentiated, possibly even separate events for signin, setup during initial 2FA, phone setup only?

Reply:
I filed a follow-up: `https://mozilla-hub.atlassian.net/browse/FXA-11968`